proposals/0000-wire-protocol.md

```diff
@@ -0,0 +1,485 @@
+
+Example of generating a discovery key in Javascript:
+
+```js
+var discoveryKey = Buffer.alloc(32)
+sodium.crypto_generichash(
+  discoveryKey, // out
+  'HYPERCORE',  // in (message)
+  publicKey     // key
+)
+```
+
+The discovery key is used in cleartext instead of the public key to avoid leaking the public key to the network; read access to hypercore feeds (including Dat archives) is controlled by limiting access to public keys.
+
+When the connection is first opened, the connecting peer sends their Feed message. The receiving peer checks that the discovery key was expected (eg, that they know of a public key matching that discovery key and are willing to synchronize the feed associated with that key). If so, they reply with their own Feed message. If not, they drop the connection.
+
+Once Feed messages are exchanged, both peers have all information they need to encrypt all further content on the channel, and do so (see below for details). The second part of the handshake is to exchange Handshake messages, which set some parameters for the channel. Handshakes also include the self-identified peer id, which can be used to detect accidental self-connections or redundant connections to the same peer (eg, over different transports). The peer id is typically random and is not authenticated.
+
+
+## Encryption Scheme
+[encryption]: #encryption
+
+After the first Feed messages are exchanged (one message in each direction, in cleartext), all further bytes exchanged over the channel are encrypted.
+
+Framing metadata (aka, message length and type) is encrypted, but a third party could likely infer message lengths (and thus potentially message types) by observing packet sizes and timing; no padding is applied at the protocol layer.
+
+The encryption scheme used is libsodium's stream primitive, specifically the XSalsa20 cipher. The cipher is fed a shared key (the primary hypercore feed public key), a nonce (selected by the sender and exchanged during handshake), and a block offset (representing all encrypted bytes sent on the connection in this direction so far).
+
+The specific libsodium function used is `crypto_stream_xsalsa20_xor_ic()`. Some interfacing code is necessary to process messages that don't align with the cipher's 64-byte chunk size; unused bytes in any particular chunk can be ignored. For example, if 1000 encrypted bytes had been sent on a connection already, and then a new 50 byte message needed to be encrypted and sent, then one would offset the message by `1000 % 64 = 40` bytes and XOR the first 24 bytes against block 15, then XOR the remaining 26 bytes against block 16. The bytes would be shifted back and recombined before sending, so only 50 bytes would go down the connection; the same process would be followed by the receiver.
```

Comment:
Is this actually what happens? I know the "unofficial" overlay implementation we have in sodium-native, which we made for hypercore calculates a block at a time (64 bytes), and then buffers whatever is left over for the next invocation, making sure that no bytes of the block go unused.

This is the API I'm talking about: https://github.com/sodium-friends/sodium-native#var-instance--crypto_stream_xor_instancenonce-key-or

Reply:
Ah, that's smart. In my implementation I just recalculated.
I've added that the unused bytes can be saved as an optimization; in general this whole paragraph was just to clarify because I got very confused when re-implementing on my own.